compatibility/bazel_tools/create-daml-app/Main.hs

```diff
@@ -167,83 +181,128 @@ main = withTempDir $ \yarnCache -> do
+                            ]
           step "Install Jest, Puppeteer and other dependencies"
           addTestDependencies "package.json" testDepsPath
-          retry 3 (callProcessSilent yarnPath ["install"])
+          patchTsDependencies uiDir "package.json"
+          -- use '--scripts-prepend-node-path' to make sure we are using the correct 'node' binary
+          retry 3 (callProcessSilent npmPath ["install", "--scripts-prepend-node-path"])
           step "Run Puppeteer end-to-end tests"
-          copyFile testTsPath (cdaDir </> "ui" </> "src" </> "index.test.ts")
-          callProcess yarnPath ["run", "test", "--ci", "--all"]
+          copyFile testTsPath (uiDir </> "src" </> "index.test.ts")
+          -- we need 'npm-cli.js' in the path for the following test
+          mbOldPath <- getEnv "PATH"
+          setEnv "PATH" (takeDirectory npmPath <> (searchPathSeparator : fromMaybe "" mbOldPath)) True
+          callProcess npmPath ["run", "test", "--ci", "--all", "--scripts-prepend-node-path"]
 
 addTestDependencies :: FilePath -> FilePath -> IO ()
 addTestDependencies packageJsonFile extraDepsFile = do
     packageJson <- readJsonFile packageJsonFile
     extraDeps <- readJsonFile extraDepsFile
     let newPackageJson = Aeson.lodashMerge packageJson extraDeps
-    Aeson.encodeFile packageJsonFile newPackageJson
-  where
-    readJsonFile :: FilePath -> IO Aeson.Value
-    readJsonFile path = do
-        -- Read file strictly to avoid lock being held when we subsequently write to it.
-        mbContent <- Aeson.decodeFileStrict' path
-        case mbContent of
-            Nothing -> fail ("Could not decode JSON object from " <> path)
-            Just val -> return val
+    BSL.writeFile packageJsonFile (Aeson.encode newPackageJson)
+
+readJsonFile :: FilePath -> IO Aeson.Value
+readJsonFile path = do
+    -- Read file strictly to avoid lock being held when we subsequently write to it.
+    content <- BSL.fromStrict <$> BS.readFile path
+    case Aeson.decode content of
+        Nothing -> error ("Could not decode JSON object from " <> path)
+        Just val -> return val
+
+extractTarGz :: FilePath -> FilePath -> IO ()
+extractTarGz targz outDir = do
+    runConduitRes
+        $ Conduit.sourceFile targz
+        .| Zlib.ungzip
+        .| Tar.untar (restoreFile (\a b -> fail (T.unpack $ a <> b)) outDir)
+
+setupNpmEnv :: FilePath -> [(TsLibrary, FilePath)] -> IO ()
+setupNpmEnv uiDir libs = do
+  forM_ libs $ \(tsLib, path) -> do
+    let name = tsLibraryName tsLib
+    let uiLibPath = uiDir </> name
+    extractTarGz path uiLibPath
+    patchTsDependencies uiDir $ uiLibPath </> "package.json"
+
+-- | Overwrite dependencies to our TypeScript libraries to point to local file dependencies in the
+-- 'ui' directory in the specified package.json file.
+patchTsDependencies :: FilePath -> FilePath -> IO ()
+patchTsDependencies uiDir packageJsonFile = do
+  packageJson0 <- readJsonFile packageJsonFile
+  case packageJson0 of
```

Comment:
I like `lens-aeson` for this kind of ad-hoc JSON manipulation but I’ll leave it up to you if you want to use it.